diff --git a/Classes/Database/RteImagesDbHook.php b/Classes/Database/RteImagesDbHook.php
index 2a1533b..5ac42bd 100644
--- a/Classes/Database/RteImagesDbHook.php
+++ b/Classes/Database/RteImagesDbHook.php
@@ -25,6 +25,7 @@
 use TYPO3\CMS\Core\Http\ApplicationType;
 use TYPO3\CMS\Core\Log\LogManager;
 use TYPO3\CMS\Core\Resource\AbstractFile;
+use TYPO3\CMS\Core\Resource\DefaultUploadFolderResolver;
 use TYPO3\CMS\Core\Resource\Exception\FileDoesNotExistException;
 use TYPO3\CMS\Core\Resource\Exception\ResourceDoesNotExistException;
 use TYPO3\CMS\Core\Resource\File;
@@ -404,7 +405,8 @@ private function modifyRteField(string $value): string
                         ) {
                             $fileName = substr(md5($absoluteUrl), 0, 10) . '.' . ($pI['extension'] ?? '');
                             // We insert this image into the user default upload folder
-                            $folder = $GLOBALS['BE_USER']->getDefaultUploadFolder();
+                            $uploadFolderResolver = GeneralUtility::makeInstance(DefaultUploadFolderResolver::class);
+                            $folder = $uploadFolderResolver->resolve($GLOBALS['BE_USER']);
                             $fileObject = $folder->createFile($fileName)->setContents($externalFile);
                             $imageConfiguration = [
                                 'width' => $attribArray['width'],
